shippingservice docs (#399)
open-telemetry/opentelemetry-demo: docs/services/shippingservice.md

# Shipping Service

This service is responsible for providing shipping information including pricing
and tracking information, when requested from Checkout Service.

> Shipping service is built primarily with Tonic, Reqwest, and opentelemtry. Other

Reviewer: opentelemtry -> OpenTelemetry libraries/components?

Author: fixed in commit

        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint(format!(
                    "{}{}",
                    env::var("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT")
                        .unwrap_or_else(|_| "http://otelcol:4317".to_string()),
                    "/v1/traces"
                )), // TODO: assume this ^ is true from config when opentelemetry crate > v0.17.0
                    // https://github.com/open-telemetry/opentelemetry-rust/pull/806 includes the environment variable.
        )
        .install_batch(opentelemetry::runtime::Tokio)
}
```

Spans and other metrics are created in this example throughout `tokio` async runtimes
found within [`tonic` server functions](https://github.com/hyperium/tonic/blob/master/examples/helloworld-tutorial.md#writing-our-server).
Be mindful of async runtime, [context guards](https://docs.rs/opentelemetry/latest/opentelemetry/struct.ContextGuard.html),
and inability to [move and clone `spans`](https://stackoverflow.com/questions/25649423/sending-trait-objects-between-threads-in-rust)
when replicating from these samples.

### Adding GRPC/HTTP instrumentation

> This service receives HTTP requests, which are instrumented in the middleware.

Reviewer 2: ```suggestion
This service receives gRPC requests
```

Is thie following true for the `shippingservice`?
```
 which are instrumented in the middleware.
```
I know that this is valid for PHP, but not sure if that applies to Rust

Author: I would consider [tonic](https://github.com/hyperium/tonic) in this case to be acting as middleware.

Reviewer 2: But it is gRPC right?

Reviewer 2: It receives gRPC and sends HTTP

> The middleware starts root span, and can pass reference in the same thread down

Reviewer 2: ```suggestion
The root span is started and passed down as reference in the same thread
```

```rust
    let tracer = global::tracer("shippingservice/get-quote");
    let mut span = tracer.start_with_context("get-quote", &parent_cx);

    span.add_event("Processing get quote request".to_string(), vec![]);

    let q = match create_quote_from_count(itemct)
        .with_context(Context::current_with_span(span))
        .await
//...create_quote_from_count...
    let f = match request_quote(count).await {
        Ok(float) => float,
        Err(err) => {
            let msg = format!("{}", err);
            return Err(tonic::Status::unknown(msg));
        }
    };

    Ok(get_active_span(|span| {
        let q = create_quote_from_float(f);
        span.add_event(
            "Received Quote".to_string(),
            vec![KeyValue::new("app.shipping.cost.total", format!("{}", q))],
        );
        span.set_attribute(KeyValue::new("app.shipping.items.count", count as i64));
        span.set_attribute(KeyValue::new("app.shipping.cost.total", format!("{}", q)));
        q
    }))
```

Note that the span cannot be used after calling an async method in the
originating function.

You may also notice the `attributes` set on the span in this example, and

> `events` propogated similarly. With any valid `span` pointer (attatched to

Reviewer: ->attached

Author: yeesh typo there thanks!

> context) the [opentelemetry api](https://docs.rs/opentelemetry/0.17.0/opentelemetry/trace/struct.SpanRef.html)

Reviewer: opentelemetry api -> OpenTelemetry API

Author: got it
